capsules/extra/src/bmm150.rs

```diff
@@ -0,0 +1,230 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2023.
+
+//! SyscallDriver for the Bosch BMM150 geomagnetic sensor.
+//!
+//! <https://www.bosch-sensortec.com/media/boschsensortec/downloads/datasheets/bst-bmm150-ds001.pdf>
+//!
+//! > The BMM150 is a standalone geomagnetic sensor for consumer
+//! > market applications. It allows measurements of the magnetic
+//! > field in three perpendicular axes. Based on Bosch’s proprietary
+//! > FlipCore technology, performance and features of BMM150 are
+//! > carefully tuned and perfectly match the demanding requirements of
+//! > all 3-axis mobile applications such as electronic compass, navigation
+//! > or augmented reality.
+//!
+//! //! Driver Semantics
+//! ----------------
+//!
+//! This driver exposes the BMM150's functionality via the [NineDof] and
+//! [NineDofClient] HIL interfaces. If gyroscope or accelerometer data is
+//! requested, the driver will return a ErrorCode.
+//!
+//! //! Usage
+//! -----
+//!
+//! ```rust
+//! # use kernel::static_init;
+//!
+//! let bmm150_i2c = static_init!(
+//!     capsules::virtual_i2c::I2CDevice,
+//!     capsules::virtual_i2c::I2CDevice::new(i2c_bus, 0x10));
+//! let bmm150 = static_init!(
+//!     capsules::bmm150::BMM150<'static>,
+//!     capsules::bmm150::BMM150::new(bmm150_i2c,
+//!         &mut capsules::BMM150::BUFFER));
+//! bmm150_i2c.set_client(bmm150);
+//! ```
+
+use core::cell::Cell;
+use kernel::hil::i2c::{self, I2CClient, I2CDevice};
+use kernel::hil::sensors::{NineDof, NineDofClient};
+use kernel::utilities::cells::{OptionalCell, TakeCell};
+use kernel::ErrorCode;
+
+#[allow(dead_code)]
+enum Registers {
+    ChipID = 0x40,
+    DATAxLsb = 0x42,
+    DATAxMsb = 0x43,
+    DATAyLsb = 0x44,
+    DATAyMsb = 0x45,
+    DATAzLsb = 0x46,
+    DATAzMsb = 0x47,
+    RHALLlsb = 0x48,
+    RHALLmsb = 0x49,
+    INTST = 0x4A,
+    CTRL1 = 0x4B,
+    CTRL2 = 0x4C,
+    CTRL3 = 0x4D,
+    CTRL4 = 0x4E,
+    LoThres = 0x4F,
+    HiThres = 0x50,
+    REPXY = 0x51,
+    REPZ = 0x52,
+}
+
+pub struct BMM150<'a, I: I2CDevice> {
+    buffer: TakeCell<'static, [u8]>,
+    i2c: &'a I,
+    ninedof_client: OptionalCell<&'a dyn NineDofClient>,
+    state: Cell<State>,
+    pending_data: Cell<bool>,
+}
+
+impl<'a, I: I2CDevice> BMM150<'a, I> {
+    pub fn new(buffer: &'static mut [u8], i2c: &'a I) -> BMM150<'a, I> {
+        BMM150 {
+            buffer: TakeCell::new(buffer),
+            i2c: i2c,
+            ninedof_client: OptionalCell::empty(),
+            state: Cell::new(State::Suspend),
+            pending_data: Cell::new(false),
```

Comment:
`state` we need for sure. The `pending_data` is just used to indicate if a reading has already been issued. I guess we can replace it by checking which state the driver is in, but having a dedicated variable to indicate if a reading is already ongoing simplifies things.

Reply:
Having two state variables generally complicates things, as any programmer updating the code needs to consider all combinations of cases. I really think it would be worth having only `state`. If additional tracking is needed it can be a value within the each state enum.